docs/design/validation.md

```diff
@@ -0,0 +1,2048 @@
+
+##### `ValueValidations` Field on `schema.Attribute`
+
+A new field that accepts a list of functions can be added to the `schema.Attribute` type. For example:
+
+```go
+schema.Attribute{
+    // ...
+    ValueValidations: []T{
+        T,
+        T,
+    },
+}
+```
+
+In this case, the framework would perform the validation similar to the previous framework `All()` wrapper function. The logical `AND` type of value validation is overwhelmingly more common in practice, which will simplify provider implementations. This still allows for an `Any()` based wrapper (logical `OR`) to be inserted if necessary.
+
+Colocating the value validation behaviors in the schema definition, means it is easier for provider developers to discover this type of validation and correlate the validation logic to the name and type information. This proposal will feel familiar to existing provider developers. New provider developers will immediately know that multiple validations are supported.
+
+##### New Attribute With Value Validation Type(s)
+
+The `schema.Attribute` type could be converted to a Go interface type and split into capabilities, similar to other interface types in the framework. For example:
+
+```go
+type Attribute interface {
+    Type(context.Context) attr.Type
+    // ...
+}
+
+type AttributeWithValueValidations struct {
+    Attribute
+    ValueValidations []T
+}
+
+// or more interfaces
+
+type AttributeWithValueValidations interface {
+    Attribute
+    ValueValidations(/* ... */) []T
+}
+```
+
+This type of proposal, in isolation, feels extraneous given the current attribute implementation. The framework does not appear to benefit from this splitting and it seems desirable that all attributes should be able to optionally enable value validation. Future considerations to allow declaring custom attribute types, outside of validation handling, are more likely to drive this type of potential change.
+
+##### Resource Level Attribute Value Validation Handling
+
+This proposal would introduce no changes to `schema.Attribute`. Instead, this would require value validation declarations at the `DataSource` and `Resource` level similar to other proposed attribute validations in the [Declaring Multiple Attribute Validation for Resources](#declaring-multiple-attribute-validation-for-resources) section.
+
+This proposal makes value validation behaviors occur at a distance, meaning it is harder for provider developers to correlate the validation logic to the name/path and type information. It would also be very verbose for even moderately sized schemas with thorough value validation. The only real potential benefit to this type of value validation is that the framework implementation is very straightforward, to just go through this single list of validations instead of walking all attributes.
```

Comment:
Sorry, missed this in the shuffle. I think what I'm trying to say is that, for example:

```go
// ignore the type signature, it's a place holder
func (r computeInstanceResource) Validate(ctx context.Context, req ValidateResourceRequest, resp *ValidateResourceResponse) {
  // assume we want to ensure, e.g., that no disks combine to more than one terabyte of storage, or something
  totalStorage := 0
  for _, disk := range plan.Disks { // not how any of this works, but pseudo-code!
    totalStorage += disk.Size
  }
  if totalStorage >= 1024*1024*1024*1024 {
    // return an error!
  }
}
```

This is a weird thing to define a reusable helper for, and it's weirder if we throw more details on this: maybe it can only be 1tb of spinning rust disks, 512gb of SSDs, and no more than 1tb of disks total. Maybe the size of the instance comes into play. That's what I mean by "incredibly complex"--I guess a better way to phrase it would be "unlikely to be reused across resources".

To implement this declaratively, you'd need to do something like this:

```go
// ignore the type signature, it's a place holder
func (r computeInstanceResource) Validators(ctx context.Context) []Validators {
  return []Validators{computeInstanceValidator{}}
}

type computeInstanceValidator struct{}

func (r computeInstanceValidator) Validate(ctx context.Context, req ValidateResourceRequest, resp *ValidateResourceResponse) {
  // assume we want to ensure, e.g., that no disks combine to more than one terabyte of storage, or something
  totalStorage := 0
  for _, disk := range plan.Disks { // not how any of this works, but pseudo-code!
    totalStorage += disk.Size
  }
  if totalStorage >= 1024*1024*1024*1024 {
    // return an error!
  }
}
```

I'm arguing this is less straightforward because you need to introduce the indirection of declaring the validation logic separately from the resource, in a reusable chunk, even though you're only using it for the resource. The code is a bit more verbose, and feels a little less intuitive, imho.